imageio/plugins/pyav.py

```diff
@@ -0,0 +1,977 @@
+            - `"FRAME"`: threads may assemble future frames
+
+
+        Yields
+        ------
+        frame : np.ndarray
+            A (decoded) video frame.
+
+
+        """
+
+        self._video_stream.thread_type = thread_type or "SLICE"
+        self._video_stream.codec_context.thread_count = thread_count
+        ffmpeg_filter = self._build_filter(filter_sequence, filter_graph)
+        ffmpeg_filter.send(None)  # init
+
+        for frame in self._container.decode(video=0):
+            frame = ffmpeg_filter.send(frame)
+
+            if frame is None:
+                continue
+
+            yield self._unpack_frame(frame, format=format)
+
+        for frame in ffmpeg_filter:
+            yield self._unpack_frame(frame, format=format)
+
+    def _unpack_frame(self, frame: av.VideoFrame, *, format: str = None) -> np.ndarray:
+        """Convert a av.VideoFrame into a ndarray
+
+        Parameters
+        ----------
+        frame : av.VideoFrame
+            The frame to unpack.
+        format : str
+            If not None, convert the frame to the given format before unpacking.
+
+        """
+
+        if format is not None:
+            frame = frame.reformat(format=format)
+
+        dtype = _format_to_dtype(frame.format)
+        shape = _get_frame_shape(frame)
+
+        planes = list()
+        for idx in range(len(frame.planes)):
+            n_channels = sum(
+                [
+                    x.bits // (dtype.itemsize * 8)
+                    for x in frame.format.components
+                    if x.plane == idx
+                ]
+            )
+            av_plane = frame.planes[idx]
+            plane_shape = (av_plane.height, av_plane.width)
+            plane_strides = (av_plane.line_size, n_channels * dtype.itemsize)
+            if n_channels > 1:
+                plane_shape += (n_channels,)
+                plane_strides += (dtype.itemsize,)
+
+            np_plane = as_strided(
+                np.frombuffer(av_plane, dtype=dtype),
+                shape=plane_shape,
+                strides=plane_strides,
+            )
+            planes.append(np_plane)
+
+        if len(planes) > 1:
+            # Note: the planes *should* exist inside a contigous memory block
+            # somewhere inside av.Frame however pyAV does not appear to expose this,
+            # so we are forced to copy the planes individually instead of wrapping
+            # them :(
+            out = np.concatenate(planes).reshape(shape)
+        else:
+            out = planes[0]
+
+        return out
+
+    def write(
+        self,
+        ndimage: Union[ArrayLike, List[ArrayLike]],
+        *,
+        codec: str,
+        is_batch: bool = True,
+        fps: int = 24,
+        in_pixel_format: str = "rgb24",
+        out_pixel_format: str = None,
+        filter_sequence: List[Tuple[str, Union[str, dict]]] = None,
+        filter_graph: Tuple[dict, List] = None,
```

Comment:
Maybe we should go easy on the type hints (e.g. using a string instead?), because this is not helping 😮  :

<img width="875" alt="image" src="https://user-images.githubusercontent.com/3015475/161930433-513f4b9a-276a-4a85-931d-d11e167157db.png">